### PR TITLE
Create docker base image for microgateway runtime

### LINKED ISSUE #842: Update the microgateway docker base image creation using the binary
### Describe your problem(s)
ATM MGW docker base image is created using the b7a base image. This can cause problems when delivering updates for the product. Hence we need to create the docker base image using the binary distribution of the microgateway
### How will you implement it

Extract the binary distribution of mgw, add proper users and groups to the image , and assign relevant permissions to paths.
---

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose
Create the docker image for the microgateway runtime.
Change the gateway script such that the config file could be overwritten by providing a command-line argument.

### Issues
Fixes #842 

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments
MacOS Mojave

---
#### Maintainers: Check before merge
- [ ] Assigned 'Type' label
- [ ] Assigned the project
- [ ] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

### CODE REVIEW

File: distribution/resources/cli-conf/toolkit-config.toml [hunk 1]
```diff
@@ -3,7 +3,7 @@ httpRequestTimeout = 1000000
 
 [token]
 baseURL = ""
-restVersion = "v1.0"
```
Comment (praminda) [inline distribution/resources/cli-conf/toolkit-config.toml]: Why do we need 0.14 which is AM 260 rest version? Even if we need to use compatibility version then this has to be 0.15 IMO

File: components/micro-gateway-cli/src/main/resources/templates/dockerConfig.mustache [hunk 2]
```diff
@@ -11,6 +11,7 @@
     debugPort:"{{containerConfig.docker.dockerConfig.debugPort}}"{{/if}}{{#if containerConfig.docker.dockerConfig.push}},
     push:{{containerConfig.docker.dockerConfig.push}}{{/if}}{{#if containerConfig.docker.dockerConfig.username}},
     username:"{{containerConfig.docker.dockerConfig.username}}"{{/if}}{{#if containerConfig.docker.dockerConfig.password}},
-    password:"{{containerConfig.docker.dockerConfig.password}}"{{/if}}
+    password:"{{containerConfig.docker.dockerConfig.password}}"{{/if}}{{#if containerConfig.docker.dockerConfig.cmd}},
+    cmd: "{{containerConfig.docker.dockerConfig.cmd}}"{{/if}}
 }
 {{/if}}
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/resources/templates/dockerConfig.mustache]:
```suggestion
{{/if}}

```

File: distribution/resources/cli-conf/toolkit-config.toml [hunk 3]
```diff
@@ -3,7 +3,7 @@ httpRequestTimeout = 1000000
...
+restVersion = "v0.14"
```
Comment (Rajith90) [inline distribution/resources/cli-conf/toolkit-config.toml]:
```suggestion
restVersion = "v1.0"
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/BuildCmd.java [hunk 4]
```diff
@@ -220,8 +220,7 @@ private void init(String projectName, String configPath, String deploymentConfig
                 copyFile.setIsBallerinaConf("true");
                 copyFile.setSource(CmdUtils.getResourceFolderLocation() + File.separator + CliConstants.GW_DIST_CONF
                         + File.separator + CliConstants.MICRO_GW_CONF_FILE);
-                copyFile.setTarget(File.separator + CliConstants.WSO2 + File.separator + CliConstants.MGW
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/BuildCmd.java]: Shall we move this docker related logic to separate method

File: distribution/resources/bin/gateway [hunk 5]
```diff
@@ -45,11 +45,13 @@ PRGDIR=`dirname "$PRG"`
...
 GW_HOME=`cd "$PRGDIR/.." ; pwd`
-export JAVA_HOME=$GW_HOME'/lib/jdk8u202-b08-jre'
 export MGW_VERSION="3.1.0"
 
 JAVA_PATH=$GW_HOME/lib/jdk8u202-b08-jre
-if [ -d "$JAVA_PATH" ]; then
+if [ -d "$JAVA_HOME" ]; then
+  echo "JAVA_HOME is externally set: $JAVA_HOME"
+else
+  echo "JAVA_HOME..."
```
Comment (praminda) [inline distribution/resources/bin/gateway]: Remove this
```suggestion
```

File: distribution/resources/bin/gateway [hunk 6]
```diff
@@ -85,8 +87,8 @@ fi
 
 # For Migwn, ensure paths are in UNIX format before anything is touched
 if $mingw ; then
-  [ -n "$JAVA_HOME" ] &&
-    JAVA_HOME="`(cd "$JAVA_HOME"; pwd)`"
+  echo $JAVA_HOME
```
Comment (praminda) [inline distribution/resources/bin/gateway]:
```suggestion
```

File: distribution/resources/bin/gateway [hunk 7]
```diff
@@ -102,6 +104,12 @@ if [ -z "$JAVACMD" ] ; then
     fi
 fi
 
+JDK_18=`$JAVACMD -version 2>&1 | grep "1.8."`
+if [ "$JDK_18" = "" ]; then
+    echo "Error: Microgateway is supported only on JDK 1.8"
```
Comment (praminda) [inline distribution/resources/bin/gateway]:
```suggestion
    echo "Error: Microgateway is supported only on Java 1.8"
```

File: distribution/resources/bin/gateway [hunk 8]
```diff
@@ -167,7 +175,7 @@ java_cmd_gateway () {
     $JAVA_OPTS \
     -Dmgw-runtime.home=${GW_HOME} \
     -Dballerina.home=${GW_HOME} \
-    -jar $args --api.usage.data.path=$GW_HOME/api-usage-data --b7a.http.accesslog.path=$GW_HOME/logs/access_logs --b7a.config.file=$GW_HOME/conf/micro-gw.conf 2>&1 | tee -a $GW_HOME/logs/microgateway.log
+    -jar ${args[0]} --api.usage.data.path=$GW_HOME/api-usage-data --b7a.http.accesslog.path=$GW_HOME/logs/access_logs --b7a.config.file=$GW_HOME/conf/micro-gw.conf ${args[@]:1}  2>&1 | tee -a $GW_HOME/logs/microgateway.log
```
Comment (praminda) [inline distribution/resources/bin/gateway]: As per the discussion we had regarding docker config loading, we don't need this change anymore

File: docker/Dockerfile [hunk 9]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (praminda) [inline docker/Dockerfile]:
```suggestion
```

File: docker/Dockerfile [hunk 10]
```diff
@@ -17,33 +18,70 @@
 FROM openjdk:jre-alpine
```
Comment (praminda) [inline docker/Dockerfile]: Use specific version of the image

File: docker/Dockerfile [hunk 11]
```diff
@@ -17,33 +18,70 @@
...
+    && chown ${USER}:${USER_GROUP} -R ${USER_HOME}  \
+    && rm -rf /root/${MGW_SERVER}.zip > /dev/null 2>&1
+
+ENV BALLERINA_HOME ${MGW_RUNTIME_HOME}/runtime
+ENV GW_HOME ${MGW_RUNTIME_HOME}
+ENV PATH $GW_HOME/bin:$PATH
+
+WORKDIR /home/ballerina
+
+USER ballerina
```
Comment (praminda) [inline docker/Dockerfile]: Use `${USER}`

Comment (VirajSalaka) [reply docker/Dockerfile]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: docker/Dockerfile (hunk 10 above)
Comment (VirajSalaka) [reply docker/Dockerfile]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: docker/Dockerfile (hunk 9 above)
Comment (VirajSalaka) [reply docker/Dockerfile]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: distribution/resources/bin/gateway (hunk 8 above)
Comment (VirajSalaka) [reply distribution/resources/bin/gateway]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: distribution/resources/bin/gateway (hunk 7 above)
Comment (VirajSalaka) [reply distribution/resources/bin/gateway]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: distribution/resources/bin/gateway (hunk 6 above)
Comment (VirajSalaka) [reply distribution/resources/bin/gateway]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: distribution/resources/bin/gateway (hunk 5 above)
Comment (VirajSalaka) [reply distribution/resources/bin/gateway]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/BuildCmd.java (hunk 4 above)
Comment (VirajSalaka) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/BuildCmd.java]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: distribution/resources/cli-conf/toolkit-config.toml (hunk 3 above)
Comment (VirajSalaka) [reply distribution/resources/cli-conf/toolkit-config.toml]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: components/micro-gateway-cli/src/main/resources/templates/dockerConfig.mustache (hunk 2 above)
Comment (VirajSalaka) [reply components/micro-gateway-cli/src/main/resources/templates/dockerConfig.mustache]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

File: distribution/resources/cli-conf/toolkit-config.toml (hunk 1 above)
Comment (VirajSalaka) [reply distribution/resources/cli-conf/toolkit-config.toml]: fixed via https://github.com/wso2/product-microgateway/pull/966/commits/fe1f68c2c9d1d39e961d2664256249498d69c5aa

**Pull request merged**